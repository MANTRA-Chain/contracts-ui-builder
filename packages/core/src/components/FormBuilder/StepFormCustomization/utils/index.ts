export * from './fieldEditorUtils';
export * from './fieldTypeUtils';
