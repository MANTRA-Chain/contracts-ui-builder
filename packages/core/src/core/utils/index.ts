export * from './button-variants';
export * from './general';
