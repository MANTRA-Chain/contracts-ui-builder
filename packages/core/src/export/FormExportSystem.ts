/**
 * FormExportSystem
 *
 * This class coordinates the form export process by integrating the
 * TemplateManager, FormCodeGenerator, and AdapterExportManager components.
 */

import { FormCodeGenerator } from './generators/FormCodeGenerator';
import { AdapterExportManager } from './AdapterExportManager';
import { PackageManager } from './PackageManager';
import { TemplateManager } from './TemplateManager';
import { ZipGenerator, type ZipProgress } from './ZipGenerator';

import type { ChainType } from '../core/types/ContractSchema';
import type { ExportOptions, ExportResult } from '../core/types/ExportTypes';
import type { BuilderFormConfig } from '../core/types/FormTypes';

/**
 * FormExportSystem class coordinates the complete form export process,
 * integrating the template system, code generator, and adapter export.
 */
export class FormExportSystem {
  private templateManager: TemplateManager;
  private formCodeGenerator: FormCodeGenerator;
  private adapterExportManager: AdapterExportManager;
  private packageManager: PackageManager;
  private zipGenerator: ZipGenerator;

  /**
   * Creates a new FormExportSystem
   */
  constructor() {
    this.templateManager = new TemplateManager();
    this.formCodeGenerator = new FormCodeGenerator();
    this.adapterExportManager = new AdapterExportManager();
    this.packageManager = new PackageManager();
    this.zipGenerator = new ZipGenerator();
  }

  /**
   * Main entry point for the export process. Exports a form based on the provided
   * form configuration, chain type, and options.
   *
   * @param formConfig Form configuration created in the builder
   * @param chainType Blockchain type (evm, solana, etc.)
   * @param functionId Function ID this form is for
   * @param options Export customization options
   * @returns An export result with the file blob and metadata
   */
  async exportForm(
    formConfig: BuilderFormConfig,
    chainType: ChainType,
    functionId: string,
    options: Partial<ExportOptions> = {}
  ): Promise<ExportResult> {
    // Ensure chainType is set in options
    const exportOptions: ExportOptions = {
      chainType,
      ...options,
    };

    try {
      // 1. Generate form component code
      const formComponentCode = this.formCodeGenerator.generateFormComponent(
        formConfig,
        chainType,
        functionId
      );

      // 2. Get adapter files if needed
      const adapterFiles =
        exportOptions.includeAdapters !== false
          ? this.adapterExportManager.getAdapterFiles(chainType)
          : {};

      // Create custom files object with generated code
      const customFiles = {
        'src/components/GeneratedForm.tsx': formComponentCode,
        ...adapterFiles,
      };

      // 3. Create the complete project using TemplateManager's createProject method
      // This will properly handle placeholders and combine template files with custom files
      const projectFiles = this.templateManager.createProject(
        exportOptions.template || 'typescript-react-vite',
        customFiles,
        exportOptions
      );

      // 4. Update package.json using PackageManager
      const originalPackageJson = projectFiles['package.json'];
      if (originalPackageJson) {
        projectFiles['package.json'] = this.packageManager.updatePackageJson(
          originalPackageJson,
          formConfig,
          chainType,
          functionId,
          exportOptions
        );
      } else {
        console.warn('No package.json found in template, using default');
        // Create a basic package.json if none exists
        projectFiles['package.json'] = JSON.stringify(
          {
            name: exportOptions.projectName || `${functionId.toLowerCase()}-form`,
            version: '0.1.0',
            private: true,
            type: 'module',
            dependencies: this.packageManager.getDependencies(formConfig, chainType),
          },
          null,
          2
        );
      }

      // Log the final project structure for debugging
      console.log('Total project files for export:', Object.keys(projectFiles).length);
      console.log('Project structure:', Object.keys(projectFiles).sort());

      // 5. Create ZIP file
      const fileName = this.generateFileName(functionId);
      const zipResult = await this.createZipFile(projectFiles, fileName, exportOptions.onProgress);

      // 6. Return the export result with dependencies from PackageManager
      return {
        zipBlob: zipResult.blob,
        fileName: zipResult.fileName,
        dependencies: this.packageManager.getDependencies(formConfig, chainType),
      };
    } catch (error) {
      console.error('Error exporting form:', error);
      throw new Error(`Export failed: ${(error as Error).message}`);
    }
  }

  /**
   * Create a ZIP file from the project files
   *
   * @param files Map of file paths to content
   * @param fileName Suggested filename for the ZIP
   * @param onProgress Optional callback for progress updates
   * @returns A result containing the ZIP blob and filename
   */
  private async createZipFile(
    files: Record<string, string>,
    fileName: string,
    onProgress?: (progress: ZipProgress) => void
  ) {
    console.log(`Creating ZIP file with ${Object.keys(files).length} files`);

    return this.zipGenerator.createZipFile(files, fileName, {
      onProgress,
      compressionLevel: 6, // Moderate compression
    });
  }

  /**
   * Generate a filename for the exported package
   * @param functionId Function ID
   * @returns A suitable filename in kebab-case format
   */
  private generateFileName(functionId: string): string {
    // Convert to kebab case by replacing non-alphanumeric chars with hyphens
    const kebabName = functionId
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-') // Replace sequences of non-alphanumeric chars with single hyphen
      .replace(/^-+|-+$/g, ''); // Remove leading/trailing hyphens
    return `${kebabName}-form.zip`;
  }
}
