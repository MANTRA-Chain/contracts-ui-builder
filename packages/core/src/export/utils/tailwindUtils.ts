import { logger } from '@openzeppelin/transaction-form-renderer';

/**
 * Modifies the content paths within tailwind.config.cjs content string.
 * Ensures that Tailwind scans all necessary files in the exported project, including
 * those from the form-renderer package, to prevent incorrect purging of utility classes.
 *
 * @param originalContent - The original content of tailwind.config.cjs.
 * @returns The modified content string.
 */
export function modifyTailwindConfigContentForExport(originalContent: string): string {
  const newContentPaths = [
    './index.html',
    './src/**/*.{js,ts,jsx,tsx}', // Scan app's src directory
    // Ensure node_modules is scanned broadly enough if form-renderer components are there
    // This path assumes form-renderer might be a direct dependency or symlinked in a way that needs explicit scanning.
    // Adjust if form-renderer styles are expected to be found elsewhere or if its components are self-contained regarding styles.
    './node_modules/@openzeppelin/transaction-form-renderer/dist/**/*.{js,ts,jsx,tsx}', // More specific path to renderer components
    './node_modules/@openzeppelin/transaction-form-renderer/src/**/*.{js,ts,jsx,tsx}', // If src is also published or linked
  ]
    .map((p) => `      '${p}'`) // Indent for formatting
    .join(',\n');

  // Regex to find 'content: [' ... ']' and replace its contents
  const modifiedContent = originalContent.replace(
    /(content\s*:\s*\[)([\s\S]*?)(\])/,
    `$1\n${newContentPaths}\n    $3` // Replace with new paths, keeping original brackets and indentation around array
  );

  if (modifiedContent === originalContent) {
    logger.warn(
      'TailwindUtils', // Changed system for logger
      'Failed to replace content paths in tailwind.config.cjs. Check config format. Using original content.'
    );
    return originalContent;
  }
  logger.info(
    'TailwindUtils',
    'Successfully modified tailwind.config.cjs content paths for export.'
  );
  return modifiedContent;
}
