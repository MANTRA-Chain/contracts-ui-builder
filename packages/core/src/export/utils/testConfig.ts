import { v4 as uuidv4 } from 'uuid';

import type { FieldType } from '@openzeppelin/transaction-form-renderer';

import type { BuilderFormConfig } from '../../core/types/FormTypes';

/**
 * Create a minimal form configuration for testing
 * @param functionName - Optional function name (defaults to 'testFunction')
 * @param chainType - Optional chain type (defaults to 'evm')
 * @returns A minimal BuilderFormConfig instance for testing
 */
export function createMinimalFormConfig(
  functionName: string = 'testFunction',
  _chainType: string = 'evm'
): BuilderFormConfig {
  return {
    functionId: functionName,
    fields: [
      {
        id: uuidv4(),
        type: 'text',
        name: 'testParam',
        label: 'Test Parameter',
        validation: {
          required: true,
        },
        helperText: `Description for Test Parameter`,
        placeholder: 'Enter test parameter',
      },
    ],
    layout: {
      columns: 1,
      spacing: 'normal',
      labelPosition: 'top',
      sections: [
        {
          id: uuidv4(),
          title: 'Test Section',
          description: 'A test section',
          fields: ['testParam'],
        },
      ],
    },
    validation: {
      mode: 'onChange',
      showErrors: 'inline',
    },
    executionConfig: {
      method: 'eoa',
      allowAny: true,
    },
    theme: {},
  };
}

/**
 * Create a complex form configuration for testing
 * @param functionName - Optional function name (defaults to 'complexFunction')
 * @param chainType - Optional chain type (defaults to 'evm')
 * @returns A complex BuilderFormConfig instance with multiple field types
 */
export function createComplexFormConfig(
  functionName: string = 'complexFunction',
  _chainType: string = 'evm'
): BuilderFormConfig {
  const basicFieldIds = [uuidv4(), uuidv4(), uuidv4()];
  const advancedFieldIds = [uuidv4(), uuidv4(), uuidv4(), uuidv4()];

  return {
    functionId: functionName,
    fields: [
      {
        id: basicFieldIds[0],
        type: 'text',
        name: 'stringParam',
        label: 'String Parameter',
        validation: { required: true },
        helperText: 'Description for String Parameter',
        placeholder: 'Enter string parameter',
      },
      {
        id: basicFieldIds[1],
        type: 'number',
        name: 'numberParam',
        label: 'Number Parameter',
        validation: { required: true, min: 0, max: 100 },
        helperText: 'Description for Number Parameter (0-100)',
        placeholder: 'Enter number parameter',
      },
      {
        id: basicFieldIds[2],
        type: 'checkbox',
        name: 'boolParam',
        label: 'Boolean Parameter',
        validation: { required: true },
        helperText: 'Description for Boolean Parameter',
      },
      {
        id: advancedFieldIds[0],
        type: 'blockchain-address',
        name: 'addressParam',
        label: 'Address Parameter',
        validation: { required: true },
        helperText: 'Description for Address Parameter',
        placeholder: 'Enter address parameter',
      },
      {
        id: uuidv4(),
        type: 'number',
        name: 'hiddenParam',
        label: 'Hidden Number',
        validation: { required: true },
        isHidden: true,
      },
      {
        id: uuidv4(),
        type: 'text',
        name: 'hardcodedTextParam',
        label: 'Fixed Text Value',
        validation: { required: true },
        isHardcoded: true,
        hardcodedValue: 'This value is fixed',
        helperText: 'This text value is set and cannot be changed.',
      },
      {
        id: uuidv4(),
        type: 'number',
        name: 'readOnlyHardcodedNumber',
        label: 'Fixed Number (Read Only)',
        validation: { required: true },
        isHardcoded: true,
        hardcodedValue: 12345,
        isReadOnly: true,
        helperText: 'This number is fixed and displayed as read-only.',
      },
      {
        id: advancedFieldIds[1],
        type: 'textarea',
        name: 'arrayParam',
        label: 'Text Array Parameter',
        validation: { required: true },
        helperText: 'Enter text array (e.g., ["a", "b"])',
        placeholder: 'e.g., ["a", "b"]',
      },
      {
        id: advancedFieldIds[2],
        type: 'textarea',
        name: 'description',
        label: 'Description',
        validation: { required: false },
        helperText: 'Optional detailed description',
        placeholder: 'Enter description...',
      },
      {
        id: advancedFieldIds[3],
        type: 'select',
        name: 'priority',
        label: 'Priority',
        validation: { required: true },
        helperText: 'Select the priority level',
        options: [
          { label: 'Low', value: 'low' },
          { label: 'Medium', value: 'medium' },
          { label: 'High', value: 'high' },
        ],
      },
    ],
    layout: {
      columns: 1,
      spacing: 'normal',
      labelPosition: 'top',
      sections: [
        {
          id: uuidv4(),
          title: 'Basic Parameters',
          description: 'Description for Basic Parameters',
          fields: [basicFieldIds[0], basicFieldIds[1], basicFieldIds[2], advancedFieldIds[0]],
        },
        {
          id: uuidv4(),
          title: 'Other Parameters',
          description: 'Description for Other Parameters',
          fields: [advancedFieldIds[1], advancedFieldIds[2], advancedFieldIds[3]],
        },
        {
          id: uuidv4(),
          title: 'Hardcoded / ReadOnly Examples',
          description: 'Fields demonstrating hardcoding and read-only features.',
          fields: ['hardcodedTextParam', 'readOnlyHardcodedNumber'],
        },
      ],
    },
    validation: {
      mode: 'onChange',
      showErrors: 'inline',
    },
    executionConfig: {
      method: 'eoa',
      allowAny: false,
      specificAddress: '0x0000000000000000000000000000000000000000',
    },
    theme: {},
  };
}

/**
 * Create a test field
 * @param type - The field type
 * @param name - The field name
 * @param label - The field label
 * @param options - Optional field options
 * @returns A field configuration object
 */
export function createTestField(
  type: FieldType,
  name: string,
  label: string,
  options: Record<string, unknown> = {}
): {
  id: string;
  type: FieldType;
  name: string;
  label: string;
  validation: { required: boolean };
  helperText: string;
  placeholder: string;
  [key: string]: unknown;
} {
  return {
    id: uuidv4(),
    type,
    name,
    label,
    validation: {
      required: true,
    },
    helperText: `Description for ${label}`,
    placeholder: `Enter ${label.toLowerCase()}`,
    ...options,
  };
}
