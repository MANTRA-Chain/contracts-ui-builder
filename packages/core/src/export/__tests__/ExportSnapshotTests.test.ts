import { describe, expect, it } from 'vitest';

import { FormExportSystem } from '../FormExportSystem';
import { createMinimalFormConfig } from '../utils/testConfig';
import { extractFilesFromZip } from '../utils/zipInspector';

import type { ChainType } from '../../core/types/ContractSchema';

describe('Export Snapshot Tests', () => {
  /**
   * Helper function to extract key files from the export for snapshot testing
   */
  async function getSnapshotFiles(chainType: ChainType = 'evm', functionName: string = 'transfer') {
    // Create the export system and form config
    const exportSystem = new FormExportSystem();
    const formConfig = createMinimalFormConfig(functionName, chainType);

    // Export the form with a consistent project name for snapshots
    const result = await exportSystem.exportForm(formConfig, chainType, functionName, {
      projectName: 'snapshot-test-project',
    });

    // Extract files from the ZIP using result.data
    expect(result.data).toBeDefined();
    const files = await extractFilesFromZip(result.data);

    // Return the files that should be snapshot tested
    return {
      packageJson: JSON.parse(files['package.json']),
      appComponent: files['src/App.tsx'],
      formComponent: files['src/components/GeneratedForm.tsx'],
      adapterIndex: files['src/adapters/index.ts'],
      [`adapter_${chainType}`]: files[`src/adapters/${chainType}/adapter.ts`],
    };
  }

  /**
   * Filter sensitive or variable content from snapshots to avoid false diffs
   */
  function prepareForSnapshot(content: string): string {
    if (!content) return '';

    // Replace timestamp comments that would cause snapshot diffs
    return (
      content
        .replace(/\/\/ Generated at:.*$/gm, '// Generated at: [timestamp]')
        .replace(/\/\* Generated:.*\*\//gm, '/* Generated: [timestamp] */')
        // Replace IDs in the form schema - more comprehensive UUID pattern
        .replace(
          /"id":\s*"[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}"/g,
          '"id": "[id]"'
        )
        // Replace field IDs directly in content - better UUID pattern matching
        .replace(
          /id: '[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}'/g,
          "id: '[id]'"
        )
        // Replace fields list in sections
        .replace(/"fields":\s*\[\s*"[a-zA-Z0-9-]+"\s*\]/g, '"fields": ["testParam"]')
        // Replace other variables that might make snapshots unstable
        .replace(/version: ["'][\d.]+["']/g, 'version: "[version]"')
    );
  }

  describe('EVM Export Snapshots', () => {
    it('should match snapshot for App component', async () => {
      const files = await getSnapshotFiles('evm');
      expect(prepareForSnapshot(files.appComponent)).toMatchSnapshot('app-component-evm');
    });

    it('should match snapshot for Form component', async () => {
      const files = await getSnapshotFiles('evm');
      expect(prepareForSnapshot(files.formComponent)).toMatchSnapshot('form-component-evm');
    });

    it('should match snapshot for adapters index', async () => {
      const files = await getSnapshotFiles('evm');
      expect(prepareForSnapshot(files.adapterIndex)).toMatchSnapshot('adapter-index-evm');
    });

    it('should match snapshot for EVM adapter', async () => {
      const files = await getSnapshotFiles('evm');
      expect(prepareForSnapshot(files.adapter_evm)).toMatchSnapshot('evm-adapter');
    });

    it('should match snapshot for package.json structure', async () => {
      const files = await getSnapshotFiles('evm');

      // Only snapshot relevant parts of package.json to avoid noise
      const { dependencies, devDependencies, scripts } = files.packageJson;
      expect({ dependencies, devDependencies, scripts }).toMatchSnapshot('package-json-evm');
    });
  });

  describe('Solana Export Snapshots', () => {
    it('should match snapshot for Solana adapter', async () => {
      const files = await getSnapshotFiles('solana');
      expect(prepareForSnapshot(files.adapter_solana)).toMatchSnapshot('solana-adapter');
    });

    it('should match snapshot for adapters index with Solana', async () => {
      const files = await getSnapshotFiles('solana');
      expect(prepareForSnapshot(files.adapterIndex)).toMatchSnapshot('adapter-index-solana');
    });

    it('should match snapshot for package.json with Solana dependencies', async () => {
      const files = await getSnapshotFiles('solana');

      // Only snapshot relevant parts of package.json to avoid noise
      const { dependencies, devDependencies, scripts } = files.packageJson;
      expect({ dependencies, devDependencies, scripts }).toMatchSnapshot('package-json-solana');
    });
  });

  describe('Cross-Chain Comparison', () => {
    it('should verify form component similarities across chains', async () => {
      // Get form components for different chains
      const evmFiles = await getSnapshotFiles('evm');
      const solanaFiles = await getSnapshotFiles('solana');

      // The core form structure should be similar
      // Looking at component definition and main return elements
      const evmFormComponent = prepareForSnapshot(evmFiles.formComponent);
      const solanaFormComponent = prepareForSnapshot(solanaFiles.formComponent);

      // Both should be React components
      expect(evmFormComponent).toContain('import');
      expect(evmFormComponent).toContain("from 'react'");
      expect(solanaFormComponent).toContain('import');
      expect(solanaFormComponent).toContain("from 'react'");

      // Both should render TransactionForm
      expect(evmFormComponent).toMatch(/return\s*\([\s\S]*?<TransactionForm/);
      expect(solanaFormComponent).toMatch(/return\s*\([\s\S]*?<TransactionForm/);

      // Both should have similar form schema structures
      expect(evmFormComponent).toMatch(/const formSchema/);
      expect(solanaFormComponent).toMatch(/const formSchema/);
    });

    it('should verify App component similarities across chains', async () => {
      // Get App components for different chains
      const evmFiles = await getSnapshotFiles('evm');
      const solanaFiles = await getSnapshotFiles('solana');

      // The App component should be nearly identical across chains
      const evmAppComponent = prepareForSnapshot(evmFiles.appComponent);
      const solanaAppComponent = prepareForSnapshot(solanaFiles.appComponent);

      // Both should import and render GeneratedForm
      expect(evmAppComponent).toMatch(/import.*GeneratedForm/);
      expect(solanaAppComponent).toMatch(/import.*GeneratedForm/);

      // Both should have similar App component structure
      expect(evmAppComponent).toMatch(/function App/);
      expect(solanaAppComponent).toMatch(/function App/);
    });
  });
});
