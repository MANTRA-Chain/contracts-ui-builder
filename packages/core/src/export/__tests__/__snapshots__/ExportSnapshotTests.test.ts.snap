// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Export Snapshot Tests > Conditional File Modifications > should modify styles.css import correctly based on isCliBuildTarget option > styles-css-cli 1`] = `
"@import 'tailwindcss' source('../../../');
@import './styles/global.css';

/* Base styles... */"
`;

exports[`Export Snapshot Tests > Conditional File Modifications > should modify styles.css import correctly based on isCliBuildTarget option > styles-css-ui 1`] = `
"@import 'tailwindcss';
@import './styles/global.css';

/* Base styles... */"
`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for App component > app-component-evm 1`] = `
"import { EvmAdapter } from '@openzeppelin/transaction-form-adapter-evm';

import GeneratedForm from './components/GeneratedForm';

interface AppProps {
  adapter: EvmAdapter;
}

/**
 * App Component
 *
 * Main application component that wraps the form.
 */
export function App({ adapter }: AppProps) {
  return (
    <div className="app">
      <header className="header">
        <h1>Transaction Form for transfer</h1>
        <p>A form for interacting with blockchain contracts</p>
      </header>

      <main className="main">
        <div className="container">
          <GeneratedForm
            adapter={adapter}
            onSubmit={(data: FormData) => {
              console.log('Transaction submitted:', data);
              // Don't return a Promise, this function should be void
            }}
          />
        </div>
      </main>

      <footer className="footer">
        <p>Generated with OpenZeppelin Transaction Form Builder</p>
        <p>Â© 2025 OpenZeppelin</p>
      </footer>
    </div>
  );
}
"
`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for EVM adapter > evm-adapter 1`] = `""`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for Form component > form-component-evm 1`] = `
"import { useState } from 'react';

import { EvmAdapter } from '@openzeppelin/transaction-form-adapter-evm';
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
  ContractStateWidget,
  TransactionForm,
  WalletConnectionProvider,
} from '@openzeppelin/transaction-form-renderer';
import type {
  ContractSchema,
  FormFieldType,
  RenderFormSchema,
  TransactionFormProps,
} from '@openzeppelin/transaction-form-types';
import { logger } from '@openzeppelin/transaction-form-utils';

// Define type for transaction result (this will be implemented in the future)
interface TransactionResult {
  txHash?: string;
  error?: string;
}

// Define props for the component, extending TransactionFormProps but making schemas optional as they are injected by the generator
interface GeneratedFormProps extends Omit<TransactionFormProps, 'schema' | 'contractSchema'> {
  adapter: EvmAdapter;
}

/**
 * Generated Transaction Form for transfer
 *
 * This component renders a form for interacting with a blockchain contract.
 * It uses the shared form-renderer package which ensures consistent behavior
 * with the preview in the form builder.
 */
export default function GeneratedForm({ onSubmit, adapter }: GeneratedFormProps) {
  const [transactionResult, setTransactionResult] = useState<TransactionResult | null>(null);
  // const [contractSchema, setContractSchema] = useState<ContractSchema | null>(null);
  const [isWidgetVisible, setIsWidgetVisible] = useState(false);
  const [loadError, _setLoadError] = useState<Error | null>(null);

  // Form schema generated from the builder and transformed by FormSchemaFactory
  const formSchema: RenderFormSchema = {
    functionId: 'transfer',
    fields: [
      {
        id: '[id]',
        type: 'text',
        name: 'testParam',
        label: 'Test Parameter',
        helperText: 'Description for Test Parameter',
        placeholder: 'Enter test parameter',
        validation: {
          required: true,
        },
      },
    ],
    layout: {
      columns: 1,
      spacing: 'normal',
      labelPosition: 'top',
    },
    validation: {
      mode: 'onChange',
      showErrors: 'inline',
    },
    theme: {},
    contractAddress: '0xe34139463bA50bD61336E0c446Bd8C0867c6fE65',
    id: 'form-transfer',
    title: 'Transfer',
    description: 'Form for interacting with the Transfer function.',
    submitButton: {
      text: 'Execute Transfer',
      loadingText: 'Processing...',
      variant: 'primary',
    },
    defaultValues: {},
  };

  // Contract schema injected by generator (loaded or uploaded by the user)
  const contractSchema: ContractSchema = {
    ecosystem: 'evm',
    name: 'MockContract',
    address: '0x1234567890123456789012345678901234567890',
    functions: [
      {
        id: 'transfer',
        name: 'transfer',
        displayName: 'Transfer',
        inputs: [],
        type: 'function',
        modifiesState: true,
      },
      {
        id: 'viewFunction',
        name: 'viewFunction',
        displayName: 'View Function',
        inputs: [],
        type: 'function',
        modifiesState: false,
        stateMutability: 'view',
      },
    ],
  };

  // Original field configurations (including hidden, hardcoded values)
  const allFieldsConfig: FormFieldType[] = [
    {
      id: '[id]',
      type: 'text',
      name: 'testParam',
      label: 'Test Parameter',
      validation: {
        required: true,
      },
      helperText: 'Description for Test Parameter',
      placeholder: 'Enter test parameter',
    },
  ];

  // Execution configuration selected in the builder
  const executionConfig: unknown | undefined = {
    method: 'eoa',
    allowAny: true,
  };
  // TODO (Export Integration): Use executionConfig at runtime to determine
  // how to sign/broadcast (e.g., standard EOA signing, Safe interaction, relayer API).

  const contractAddress = formSchema.contractAddress;

  // TODO: Enable this useEffect as a fallback?
  // If the adapter supports runtime schema loading (e.g., via Etherscan)
  // and the injected schema is missing or invalid, this could attempt to load it.
  /*
  useEffect(() => {
    setLoadError(null);
    setContractSchema(null);

    if (contractAddress) {
      adapter
        .loadContract(contractAddress)
        .then(setContractSchema)
        .catch((err: unknown) => {
          // Catch error during contract loading
          logger.error('GeneratedForm', 'Error loading contract schema:', err);
          // Create a new Error object if caught value is not already one
          const errorToSet =
            err instanceof Error ? err : new Error('Failed to load contract state');
          setLoadError(errorToSet);
          setContractSchema(null);
        });
    } else {
      setContractSchema(null);
    }
  }, [contractAddress, adapter]);
  */

  // Decide which schema to use: prioritize injected, fallback maybe later?
  const schemaToUse = contractSchema; // Sticking to injected schema for now

  const toggleWidget = () => {
    setIsWidgetVisible((prev: boolean) => !prev);
  };

  // Handle form submission
  const handleSubmit = (formData: FormData) => {
    // Log the execution config (will be used for signing/broadcasting logic later)
    console.log('Using Execution Config:', executionConfig);

    // Convert FormData to Record<string, unknown> for adapter
    const submittedInputs: Record<string, unknown> = {};
    formData.forEach((value, key) => {
      submittedInputs[key] = value;
    });

    if (!schemaToUse) {
      console.error('handleSubmit: Contract schema is not available!');
      setTransactionResult({ error: 'Contract schema not loaded.' });
      return;
    }

    try {
      const functionId = 'transfer';
      const formattedData = adapter.formatTransactionData(
        schemaToUse,
        functionId,
        submittedInputs,
        allFieldsConfig
      );

      // --- Integration with onSubmit prop ---
      if (onSubmit) {
        // We might need to reconsider passing raw FormData if onSubmit expects the formatted data
        // For now, assume onSubmit handles the interaction
        onSubmit(formData);
        setTransactionResult({ txHash: '0x_SUBMITTED_VIA_PROP' }); // Indicate submission via prop
      } else {
        // --- Default submission logic (if no onSubmit provided) ---
        // This would typically involve signing and broadcasting
        console.log('Formatted data:', formattedData);
        // const result = await adapter.signAndBroadcast(formattedData);
        // setTransactionResult(result);
        // For template testing:
        setTransactionResult({ txHash: '0x_MOCK_TX_HASH' });
        logger.info('GeneratedForm', 'Mock submission successful!');
      }
    } catch (error) {
      logger.error('GeneratedForm', 'Submission error:', error);
      // TODO: Set an error state to display to the user
      setTransactionResult({ error: (error as Error).message });
    }
  };

  return (
    <WalletConnectionProvider adapter={adapter}>
      <div className="flex gap-4">
        <div className="flex-1">
          <Card>
            <CardHeader>
              {/* Render title unconditionally; React handles empty strings */}
              <CardTitle>{/*@@formSchema.title@@*/}</CardTitle>
              {/* Render description unconditionally */}
              <CardDescription>{/*@@formSchema.description@@*/}</CardDescription>
            </CardHeader>
            <CardContent className="space-y-4">
              {transactionResult && (
                <div className="transaction-result rounded-md bg-green-50 p-4 text-green-800">
                  <h3 className="font-medium">Transaction Successful!</h3>
                  <p className="mt-2 text-sm">
                    Transaction Hash: {transactionResult.txHash || 'N/A'}
                  </p>
                </div>
              )}
              {/* Check the actual contractAddress variable at runtime */}
              {contractAddress ? (
                <TransactionForm
                  schema={formSchema}
                  contractSchema={schemaToUse}
                  adapter={adapter}
                  onSubmit={handleSubmit}
                />
              ) : (
                <div className="text-destructive-foreground rounded-md bg-destructive p-4">
                  <h3 className="font-medium">Configuration Error</h3>
                  <p className="mt-2 text-sm">Missing contract address in the form schema.</p>
                </div>
              )}
            </CardContent>
          </Card>
        </div>

        {contractAddress && (
          <div className="w-[300px] flex-shrink-0">
            <div className="sticky top-4">
              <ContractStateWidget
                contractSchema={schemaToUse}
                contractAddress={contractAddress}
                adapter={adapter}
                isVisible={isWidgetVisible}
                onToggle={toggleWidget}
                error={loadError}
              />
            </div>
          </div>
        )}
      </div>
    </WalletConnectionProvider>
  );
}
"
`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for adapters index > adapter-index-evm 1`] = `""`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for package.json structure > package-json-evm 1`] = `
{
  "dependencies": {
    "@openzeppelin/transaction-form-adapter-evm": "^0.0.1",
    "@openzeppelin/transaction-form-renderer": "^1.0.0",
    "@openzeppelin/transaction-form-types": "^0.0.1",
    "@openzeppelin/transaction-form-utils": "^0.1.0",
    "lodash": "^4.17.21",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
    "react-hook-form": "^7.54.2",
    "viem": "^2.28.0",
    "wagmi": "^2.15.0",
  },
  "devDependencies": {
    "@types/lodash": "^4.17.16",
    "@types/react": "^19.0.10",
    "@types/react-dom": "^19.0.3",
    "@typescript-eslint/eslint-plugin": "^8.26.0",
    "@typescript-eslint/parser": "^8.26.0",
    "@vitejs/plugin-react": "^4.3.1",
    "autoprefixer": "^10.4.17",
    "eslint": "^9.3.0",
    "eslint-plugin-react-hooks": "^5.0.0",
    "eslint-plugin-react-refresh": "^0.4.5",
    "postcss": "^8.4.35",
    "tailwindcss": "^4.0.0",
    "typescript": "^5.8.2",
    "vite": "^6.2.5",
  },
  "scripts": {
    "build": "tsc && vite build",
    "check-deps": "npm outdated",
    "dev": "vite",
    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
    "preview": "vite preview",
    "update-form-renderer": "npm update @openzeppelin/transaction-form-renderer",
  },
}
`;

exports[`Export Snapshot Tests > Solana Export Snapshots > should match snapshot for Solana adapter > solana-adapter 1`] = `""`;

exports[`Export Snapshot Tests > Solana Export Snapshots > should match snapshot for adapters index with Solana > adapter-index-solana 1`] = `""`;

exports[`Export Snapshot Tests > Solana Export Snapshots > should match snapshot for package.json with Solana dependencies > package-json-solana 1`] = `
{
  "dependencies": {
    "@openzeppelin/transaction-form-adapter-evm": "^0.0.1",
    "@openzeppelin/transaction-form-renderer": "^1.0.0",
    "@openzeppelin/transaction-form-types": "^0.0.1",
    "@openzeppelin/transaction-form-utils": "^0.1.0",
    "lodash": "^4.17.21",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
    "react-hook-form": "^7.54.2",
    "viem": "^2.28.0",
    "wagmi": "^2.15.0",
  },
  "devDependencies": {
    "@types/lodash": "^4.17.16",
    "@types/react": "^19.0.10",
    "@types/react-dom": "^19.0.3",
    "@typescript-eslint/eslint-plugin": "^8.26.0",
    "@typescript-eslint/parser": "^8.26.0",
    "@vitejs/plugin-react": "^4.3.1",
    "autoprefixer": "^10.4.17",
    "eslint": "^9.3.0",
    "eslint-plugin-react-hooks": "^5.0.0",
    "eslint-plugin-react-refresh": "^0.4.5",
    "postcss": "^8.4.35",
    "tailwindcss": "^4.0.0",
    "typescript": "^5.8.2",
    "vite": "^6.2.5",
  },
  "scripts": {
    "build": "tsc && vite build",
    "check-deps": "npm outdated",
    "dev": "vite",
    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
    "preview": "vite preview",
    "update-form-renderer": "npm update @openzeppelin/transaction-form-renderer",
  },
}
`;
